scapy/layers/ipsec.py

```diff
@@ -127,8 +129,47 @@ class ESP(Packet):
 
 bind_layers(IP, ESP, proto=socket.IPPROTO_ESP)
 bind_layers(IPv6, ESP, nh=socket.IPPROTO_ESP)
-bind_layers(UDP, ESP, dport=4500)  # NAT-Traversal encapsulation
-bind_layers(UDP, ESP, sport=4500)  # NAT-Traversal encapsulation
+
+
+class NAT_T(Packet):
+    """
+    UDP Encapsulation of IPsec ESP Packets [RFC3948] (for NAT-Traversal)
+    """
+    name = 'NAT_T'
+
+    __slots__ = [
+        "_non_esp"
+    ]
+
+    fields_desc = [
+        ConditionalField(XIntField("non_esp", 0x0), lambda pkt: pkt._non_esp)
+    ]
+
+    def __init__(self, *args, **kwargs):
+        self._non_esp = False
+        super(NAT_T, self).__init__(*args, **kwargs)
+
+    def pre_dissect(self, s):
+        # type: (bytes) -> bytes
+        if isinstance(s, bytes) and len(s) >= 4:
+            self._non_esp = struct.unpack("!I", s[0:4])[0] == 0x00
+        else:
+            self._non_esp = False
+        return s
+
+    def add_payload(self, payload):
+        # type: (Union[Packet, bytes]) -> None
+        super(NAT_T, self).add_payload(payload)
+        self._non_esp = isinstance(self.payload, ISAKMP_class)
```

Comment:
I would have preferred to just test `isinstance(pkt.payload, ISAKMP)` as you did previously, instead of this

Reply:
Originally, the `IKEv2` class was unrelated to `ISAKMP`, it was derived directly from `Packet`. So I concidered two variants:

- Derive `IKEv2` from `ISAKMP`.
- Check for `isinstance(self.payload, ISAKMP) or isinstance(self.payload, IKEv2)` instead of introducing a class inheritance

The first variant is nicer and more object oriented. However, I hesitated to introduce direct inheritance because both classes have a very similar implementation (in fact, `ISAKMP` was copy&pasted&renamed to `IKEv2`) using the same methods and similar payloads, I feared that their implementations would interfere in a subtle way. When I discovered `ISAKMP_class`, I thought it might be less risky to use it as common base class.

But I can give variant one another try.

Reply:
What would be your preferred choice?

Reply:
Given the original class hierarchy it seemed to me the natural choice to derive `IKEv2_class` from `ISAKMP_class`, not `IKEv2` from `ISAKMP`.
```
         Packet                                 Packet                
            |                                     |                   
            |                                     |                   
      ISAKMP_class                           IKEv2_class              
       /       \                               /     \                
      /         \                             /       \               
  ISAKMP       ISAKMP_payload_*            IKEv2      IKEv2_payload_*
```
On second thought I begin to seriously doubt whether it was a good idea to introduce any inheritance relationship between lhs and rhs at all. It would only make sense with double inheritances and nobody wants such thing.

I'll revert this change and stick to the double `isinstance(self.payload, ISAKMP) or isinstance(self.payload, IKEv2)` check instead.

Reply:
See 09d0590f48a52f1f950f4239c0caee61beba1f27.